### PR TITLE
Add NPOS for collators 

### PR DESCRIPTION
**Pull request by c410-f3r**
cc #78

Tested locally. Parachains are collecting blocks and receiving rewards
**Sometimes** the parachain selection algorithm is not picking an collator and this issue still needs to be figure out. Not sure if it is related to a minimum set of collators.

Capture d’écran de 2021-05-03 23-44-58

### CODE REVIEW

File: primitives/src/ztg.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// Distribution
```
Comment (c410-f3r) [inline primitives/src/ztg.rs:1]: Taken from the website.

The only necessary parameters here are `PARACHAIN_LEASE`, `CROWDLOAN`, `LIQUIDITY_MINING` and `STAKING`. Everything else is surplus but will eventually be used in the future.

File: primitives/src/ztg.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+/// Total ZTG amount for airdrops
+pub const AIRDROPS: u128 = 3_000_000;
+
+/// Total ZTG amount for collators
+pub const COLLATORS: u128 = 10_000_000;
```
Comment (c410-f3r) [inline primitives/src/ztg.rs:7]: This was invented with a random number and needs to be modified. It is also necessary to specify an amount of ZTG for liquidity mining.

File: node/src/chain_spec/dev.rs [hunk 3]
```diff
@@ -13,16 +15,34 @@ pub fn dev_config(
...
+                        (
+                            get_account_id_from_seed::<sr25519::Public>("Alice//stash"),
+                            None,
+                            1_000 * zeitgeist_primitives::constants::BASE,
+                        ),
+                        (
+                            get_account_id_from_seed::<sr25519::Public>("Bob//stash"),
+                            None,
+                            1_000 * zeitgeist_primitives::constants::BASE,
+                        ),
```
Comment (lsaether) [inline node/src/chain_spec/dev.rs:31]: Do we need Bob as a staker here? below Bob is not set as an authority

File: node/src/chain_spec/mod.rs [hunk 4]
```diff
@@ -1,15 +1,41 @@
...
+    parachain_staking::InflationInfo {
+        annual: parachain_staking::Range {
+            ideal: Perbill::from_percent(ztg::STAKING),
+            max: Perbill::from_percent(ztg::STAKING),
+            min: Perbill::from_percent(ztg::STAKING),
+        },
+        expect: parachain_staking::Range {
+            ideal: ztg::COLLATORS * zeitgeist_primitives::constants::BASE,
+            min: ztg::COLLATORS * zeitgeist_primitives::constants::BASE,
+            max: ztg::COLLATORS * zeitgeist_primitives::constants::BASE,
```
Comment (lsaether) [inline node/src/chain_spec/mod.rs:29]: I haven't had the chance to delve in the staking pallet yet, so I'm not yet sure what configuring `max` and `min` means. What happens if less or more than `ztg::COLLATORS * zeitgeist_primitives::constants::BASE` is staked?

**Review by lsaether — APPROVED** [review]
Looks good overall! just had a couple comments

File: node/src/chain_spec/dev.rs (hunk 3 above)
Comment (c410-f3r) [reply node/src/chain_spec/dev.rs:31]: Aura and Grandpa authorities are for standalone nodes and these `Alice//stash` and `Bob//stash` accounts are for an initial set of stakers for testing.

Comment (lsaether) [reply node/src/chain_spec/dev.rs:31]: 🤦 d'oh... you're right

File: node/src/chain_spec/mod.rs (hunk 4 above)
Comment (c410-f3r) [reply node/src/chain_spec/mod.rs:29]: Yes, it is confusing. `expect` is the amount of ZTG that will flow to collators with no less than `min` and no more than `max`, `annual` is self-descriptive and `round`  is the round inflation percentage defined in the `runtime`. Kind of redundant.
MB received a grant and these crates are still in active development so things will probably change in the near future.

Additional resources:

* https://github.com/PureStake/moonbeam/blob/d40f492f7d6ba6debd529da4179821d9a8d6d110/pallets/parachain-staking/src/inflation.rs#L98-L105.
* https://github.com/PureStake/moonbeam/blob/e73891ff1125e205e2f879f9a668b97a0b7fda07/node/src/chain_spec.rs#L207-L226

**Comment by c410-f3r** [comment]
Updated inflation information to better clarify its intentions

File: scripts/parachain/local.sh [hunk 5]
```diff
@@ -27,7 +27,7 @@ cargo build --features parachain --manifest-path ../../node/Cargo.toml --release
 ../release/zeitgeist export-genesis-wasm --chain $CHAIN > para-wasm
 
 ./target/release/polkadot build-spec --chain rococo-local --disable-default-bootnode > rococo-local-plain.json
-echo "s/\"paras\": \[\]/\"paras\": [[9123, { \"genesis_head\": \"$(cat para-genesis)\", \"validation_code\": \"$(cat para-wasm)\", \"parachain\": true }]]/" > sed.txt
+echo "s/\"paras\": \[\]/\"paras\": [[$PARACHAIN_ID, { \"genesis_head\": \"$(cat para-genesis)\", \"validation_code\": \"$(cat para-wasm)\", \"parachain\": true }]]/" > sed.txt
```
Comment (sea212) [inline scripts/parachain/local.sh:30]: In combination with `set -x`, this spams the console with the wasm blob.

One solution:
```suggestion
set +x
SET_MSG="+ echo '\"s/\"paras\": \[\]/\"paras\": [[$PARACHAIN_ID, { \"genesis_head\": \"$(cat para-genesis)\", \"validation_code\": \"--snip--\", \"parachain\": true }]]/\"' > sed.txt"
echo "$SET_MSG"
echo "s/\"paras\": \[\]/\"paras\": [[$PARACHAIN_ID, { \"genesis_head\": \"$(cat para-genesis)\", \"validation_code\": \"$(cat para-wasm)\", \"parachain\": true }]]/" > sed.txt
set -x
```

File: scripts/parachain/local.sh [hunk 6]
```diff
@@ -46,27 +46,60 @@ start_validator() {
...
+    -lruntime=trace
+}
+
+# Feel free to comment, add or remove collators. Just remember that #Validators > #Collators
+
+start_collator 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY 30333 9933 9944 32000 8200 9200 & node_pid=$!
+start_collator 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty 30334 9934 9945 32001 8201 9201 & node_pid=$!
+start_collator 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY 30335 9935 9946 32002 8202 9202 & node_pid=$!
+start_collator 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty 30336 9936 9947 32003 8203 9203 & node_pid=$!
+start_collator 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty 30337 9937 9948 32004 8204 9204
```
Comment (sea212) [inline scripts/parachain/local.sh:105]: All the collators and validators spam the same console. I would prefer if only one validator and one collator write output to the console.

**Review by sea212 — COMMENTED** [review]
Thanks for implementing that extensive core feature.
The PR looks good overall. I can run a local rococo parachain using the supplied script. It leaves a great feeling seeing how our chain interacts with the relay chain.
I can also run a local testnet using this command: `cargo run --features parachain --release -- --chain local --tmp`
However, executing this command for the *dev* chain leaves me with that error:
> Error: Input("Error parsing spec file: missing field `parachain_id` at line 48 column 1")
And executing it using the *battery_park* chain leaves me with that error:
> Error: Input("Error parsing spec file: missing field `parachain_id` at line 48 column 1")

File: node/src/rpc.rs [hunk 7]
```diff
@@ -1,11 +1,4 @@
-//! A collection of node-specific RPC methods.
-//! Substrate provides the `sc-rpc` crate, which defines the core RPC layer
-//! used by Substrate nodes. This file extends those RPC definitions with
-//! capabilities that are specific to this project's runtime configuration.
-
-#![warn(missing_docs)]
-
-pub use sc_rpc_api::DenyUnsafe;
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply node/src/rpc.rs:8]: `DenyUnsafe` re-export is not used anywhere, #![warn(missing_docs)] should be used in the root module or in the `RUST_FLAGS` environment variable, this module is only used by the standalone module and the contents are self-describing  IMO.

File: scripts/tests/misc.sh [hunk 8]
```diff
@@ -14,7 +14,5 @@ test_package_with_feature primitives std
 for package in zrml/*
 do
   test_package_with_feature "$package" std
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply scripts/tests/misc.sh:16]: Thanks

File: scripts/parachain/local.sh (hunk 6 above)
Comment (c410-f3r) [reply scripts/parachain/local.sh:105]: I am open to suggestion. How to output one collator and one validator and keep everything else running in the background? It is also important to close all processes if the scripts exists

File: primitives/src/constants.rs [hunk 9]
```diff
@@ -40,3 +43,8 @@ parameter_types! {
     pub const MinWeight: Balance = BASE;
     pub const SwapsPalletId: PalletId = PalletId(*b"zge/swap");
 }
+
+// Collator staking
+parameter_types! {
+    pub const DefaultBlocksPerRound: u32 = 2 * BLOCKS_PER_MINUTE as u32;
+}
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply primitives/src/constants.rs:50]: This particular parameter is used by `runtime` and `node`, thus the decentralized location.
Ok, I will move everything related to make it concise with the rest

File: scripts/parachain/local.sh (hunk 6 above)
Comment (c410-f3r) [reply scripts/parachain/local.sh:105]: NVM. Already figured it out

**Comment by c410-f3r** [comment]
@lsaether @sea212 Thanks for the review and feel free to suggest more things if necessary

**Comment by sea212** [comment]
> @lsaether @sea212 Thanks for the review and feel free to suggest more things if necessary

Sure, thanks for incorporating the suggestions. One more thing, can you investigate into this please:

> I can also run a local testnet using this command: `cargo run --features parachain --release -- --chain local --tmp`
> However, executing this command for the _dev_ chain leaves me with that error:
> 
> > Error: Input("Relay chain argument error: Invalid input: Error opening spec file: No such file or directory (os error 2)")
> 
> And executing it using the _battery_park_ chain leaves me with that error:
> 
> > Error: Input("Error parsing spec file: missing field `parachain_id` at line 48 column 1")

**Comment by c410-f3r** [comment]
> > @lsaether @sea212 Thanks for the review and feel free to suggest more things if necessary
> 
> @c410-f3r Sure, thanks for incorporating the suggestions. One more thing, can you investigate into this please:
> 
> > I can also run a local testnet using this command: `cargo run --features parachain --release -- --chain local --tmp`
> > However, executing this command for the _dev_ chain leaves me with that error:
> > > Error: Input("Relay chain argument error: Invalid input: Error opening spec file: No such file or directory (os error 2)")
> > 
> > 
> > And executing it using the _battery_park_ chain leaves me with that error:
> > > Error: Input("Error parsing spec file: missing field `parachain_id` at line 48 column 1")

The automatic relay-chain specification for `dev` was `rococo-dev` but this spec doesn't exist, thus the error. To workaround:
```
cargo run --features parachain --release -- --chain dev --tmp -- --chain rococo-local
```
In any case, this issue was already fixed so now  `cargo run --features parachain --release -- --chain dev --tmp` will work

**Comment by c410-f3r** [comment]
The `node/res/bp_parachain.json` state had to be changed because new initial staking parameters were inserted into the runtime and node.

**Comment by c410-f3r** [comment]
@sea212 Can you take a look again? All your comments were addressed

**Review by lsaether — APPROVED** [review]

**Review by sea212 — APPROVED** [review]
Looks good now, thanks for including all suggestions.

**Comment by c410-f3r** [comment]
Thanks!

**Pull request merged**